File: topics/COUNT_MIN_SKETCH/count_min_sketch.py
import hashlib
import math

class CountMinSketch:
    """
    A probabilistic data structure that serves as a frequency table of events in a stream of data.
    """
    def __init__(self, epsilon, delta):
        """
        Initialize the Count-Min Sketch.
        epsilon : float : Error margin (e.g., 0.01 for 1% error)
        delta : float : Probability of exceeding the error margin (e.g., 0.05 for 5% chance)
        """
        # Calculate width (w) and depth (d) based on desired error bounds
        self.width = math.ceil(math.e / epsilon)
        self.depth = math.ceil(math.log(1.0 / delta))

        # Initialize the 2D array of counters with zeros
        self.table = [[0] * self.width for _ in range(self.depth)]

    def _hash(self, item, seed):
        """Generates a column index for a specific row (seed)."""
        # Create a unique hash for the item+seed combination
        h = hashlib.sha256((str(item) + str(seed)).encode('utf-8')).hexdigest()
        # Modulo the width to get the column index
        return int(h, 16) % self.width

    def add(self, item, count=1):
        """
        Add an item to the sketch, incrementing its frequency by 'count'.
        """
        # Iterate over every row (depth)
        for i in range(self.depth):
            # Calculate the column index using the row number as a seed
            col = self._hash(item, i)
            # Increment the counter
            self.table[i][col] += count

    def get_count(self, item):
        """
        Estimate the frequency of an item. Returns the minimum count across all hash rows.
        """
        min_count = float('inf')
        for i in range(self.depth):
            col = self._hash(item, i)
            # Find the minimum counter value across all rows for this item
            min_count = min(min_count, self.table[i][col])
        return min_count

if __name__ == '__main__':
    # Sample Input: 1% error margin, 5% failure probability
    cms = CountMinSketch(epsilon=0.01, delta=0.05)
    print(f"CMS initialized with Depth (hash functions): {cms.depth}, Width (buckets): {cms.width}")

    # Add items to the stream
    stream = ['apple'] * 100 + ['banana'] * 50 + ['orange'] * 10

    print("Processing stream...")
    for item in stream:
        cms.add(item)

    # Also add some random noise that might cause hash collisions
    for i in range(500):
        cms.add(f"noise_{i}")

    # Sample Output
    print("\nEstimated Frequencies:")
    print(f"Apple (Actual 100): {cms.get_count('apple')}")
    print(f"Banana (Actual 50): {cms.get_count('banana')}")
    print(f"Orange (Actual 10): {cms.get_count('orange')}")
    print(f"Grape (Actual 0): {cms.get_count('grape')}")

File: topics/COUNT_MIN_SKETCH/count_min_sketch.rs
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

/// Count-Min Sketch for probabilistic frequency estimation
pub struct CountMinSketch {
    /// The 2D array of counters (flattened into 1D for efficiency)
    table: Vec<u32>,
    /// Number of columns (w)
    width: usize,
    /// Number of rows/hash functions (d)
    depth: usize,
}

impl CountMinSketch {
    /// Initialize with epsilon (error margin) and delta (probability of exceeding error)
    pub fn new(epsilon: f64, delta: f64) -> Self {
        // Calculate width: w = ceil(e / epsilon)
        let width = (std::f64::consts::E / epsilon).ceil() as usize;
        // Calculate depth: d = ceil(ln(1 / delta))
        let depth = (1.0 / delta).ln().ceil() as usize;

        CountMinSketch {
            table: vec![0; width * depth],
            width,
            depth,
        }
    }

    /// Helper to get the column index for a given item and row (seed)
    fn get_hash<T: Hash>(&self, item: &T, seed: usize) -> usize {
        let mut hasher = DefaultHasher::new();
        // Hash the item
        item.hash(&mut hasher);
        // Salt the hash with the row index (seed) to simulate independent hash functions
        seed.hash(&mut hasher);

        let hash_val = hasher.finish();
        (hash_val % (self.width as u64)) as usize
    }

    /// Add an item to the sketch, incrementing by 'count'
    pub fn add<T: Hash>(&mut self, item: &T, count: u32) {
        for i in 0..self.depth {
            let col = self.get_hash(item, i);
            // Calculate 1D index from 2D coordinates: (row * width) + col
            let index = i * self.width + col;
            self.table[index] += count;
        }
    }

    /// Estimate the frequency of an item
    pub fn get_count<T: Hash>(&self, item: &T) -> u32 {
        let mut min_count = u32::MAX;
        for i in 0..self.depth {
            let col = self.get_hash(item, i);
            let index = i * self.width + col;
            let current_count = self.table[index];

            // The true frequency is bounded by the minimum counter
            if current_count < min_count {
                min_count = current_count;
            }
        }
        min_count
    }
}

fn main() {
    // Sample Input: 1% error margin, 5% failure probability
    let mut cms = CountMinSketch::new(0.01, 0.05);
    println!("CMS initialized with Depth: {}, Width: {}", cms.depth, cms.width);

    println!("Processing stream...");
    // Add known items
    for _ in 0..100 { cms.add(&"apple", 1); }
    for _ in 0..50 { cms.add(&"banana", 1); }
    for _ in 0..10 { cms.add(&"orange", 1); }

    // Add noise to simulate a real stream and force collisions
    for i in 0..500 {
        cms.add(&format!("noise_{}", i), 1);
    }

    // Sample Output
    println!("\nEstimated Frequencies:");
    println!("Apple (Actual 100): {}", cms.get_count(&"apple"));
    println!("Banana (Actual 50): {}", cms.get_count(&"banana"));
    println!("Orange (Actual 10): {}", cms.get_count(&"orange"));
    println!("Grape (Actual 0): {}", cms.get_count(&"grape"));
}
